README.md
# Dumb Lua Parser

*Dumb Lua Parser* is a library for tokenizing Lua code or creating ASTs (Abstract Syntax Trees) and converting the data back to Lua.
It can also optimize and minify code.

The library is a [single Lua file](dumbParser.lua) with no dependencies other than Lua itself.
It works with Lua 5.1, 5.2, 5.3 and 5.4. LuaJIT should work too.
MIT license.

Download the latest stable release [here](https://github.com/ReFreezed/DumbLuaParser/releases/latest)
or, if you're cloning the repository, checkout the last commit with a version tag.
Also available on [LuaRocks](https://luarocks.org/modules/refreezed/dumbluaparser).


## Basic usage

```lua
local parser = require("dumbParser")

local tokens = parser.tokenizeFile("cool.lua")
local ast    = parser.parse(tokens)

parser.simplify(ast)
parser.printTree(ast)

local lua = parser.toLua(ast, true)
print(lua)
```

Check out [examples.lua](examples.lua) for some examples.


## Documentation

See the [website](http://refreezed.com/luaparser/docs/) or the top of the [source file](dumbParser.lua).



CHANGELOG.txt
Changelog
Dumb Lua Parser

v2.3 (2022-06-23)
- Added utility functions: findDeclaredNames(), findGlobalReferences(), findShadows().
- Added token functions: updateToken(), cloneToken().
- simplify() handles some more cases.
- Fixed newToken("string") always raising an error.

v2.2 (2022-06-02)
- Added function: isStatement().
- Fixed updateReferences() not always updating all references properly (which sometimes broke minify()).
- Fixed backslashes in string literals not being outputted as \\.
- LuaJIT: Fixed validateTree() reporting 64-bit integers as invalid literal values.

v2.1 (2021-09-03)
- Added AST functions: parseExpression(), valueToAst(), validateTree(), isExpression(), newNodeFast().
- Added AST node fields: AstNode.pretty, AstNode.prefix, AstNode.suffix .
- tokenize(): Added option to keep whitespace tokens (which is a new token type).
- toLua(): Added nodeCallback argument.
- toLua(): Outputting less parentheses around some binary expressions.
- toLua(): Fixed invalid output of immediately invoked function expressions and similar expressions.
- toLua(): Fixed "." not always being always being considered when inserting spaces (e.g. for `x .. ...`).
- TableField.key is now optional if TableField.generatedKey is true.
- Fixed "_ENV" getting minified in Lua 5.2+.
- Fixed parsing error for chunks ending in return statements without values.

v2.0 (2021-06-19)
Major changes:
- Added support for Lua 5.4 and LuaJIT.
- Added functions: simplify(), optimize(), minify().
- Tokens are now represented by individual tables. Token streams are now simple arrays of tokens.
- Changed AstIdentifier.declaration to point to the declared identifier instead of the parent declaration/function/'for' loop.
- Changed AstLabel.name and AstGoto.name to be strings instead of identifiers.
- Removed AstFunction.vararg. (Any vararg parameter is now at the end of AstFunction.parameters.)
- Parsing and tokenizing functions no longer print anything on error - they only return the message (which now includes more information).
- toLua() no longer prints anything on error - instead it returns the message.
Added:
- Added support for binary number notation '0b1101001' (from LuaJIT).
- Added special parsing rules for 'n/0'.
- Added AST functions: cloneNode(), cloneTree(), traverseTreeReverse(), getChild(), setChild(), addChild(), removeChild().
- Added token functions: concatTokens(), newToken().
- Added function: formatMessage().
- Added AST node fields: AstVararg.declaration, AstGoto.label, AstIdentifier.attribute .
- Added more common fields to AST nodes.
- Added argument 'leavesFirst' to traverseTree().
- Added argument 'attributeName' to newNode("identifier").
- Added argument 'updateTopNodePosition' to updateReferences().
- Added parser settings: printIds, printLocations, indentation, constantNameReplacementStringMaxLength.
- Added parser constants: INT_SIZE, MAX_INT, MIN_INT.
- Added internal fallback parser for string values so features added in later Lua versions are supported in all versions.
- LuaJIT: Support numeral suffixes 'LL', 'ULL' and 'i'.
Changed:
- 'break' statements are allowed in the middle of blocks (according to Lua 5.2).
- '-0' is no longer normalized to '0' in Lua 5.2 and earlier (except by simplify()) to match Lua's behavior.
- newNode("label") and newNode("goto") now require a 'labelName' argument.
- Renamed parse(path) function variant to parseFile().
- The path argument for parse(luaString) function variant is now optional.
- Renamed tokenizeString() to tokenize() (to follow the same naming scheme as parse() and parseFile()).
- toLua(): Improved string literal output.
- toLua(): Integers gets outputted as integers (as opposed to floats with scientific notation) in more cases.
- Updated "malformed number" error detection.
- Parsing: Improved location information for blocks and functions.
- Whitespace is no longer trimmed from comment tokens.
- Improved error messages.
Removed:
- Removed functions: newTokenStream(), insertToken(), removeToken().
Fixed:
- Fixed toLua() outputting expressions like '- -x' as '--x'.
- Fixed printTokens() and printNode() printing unprintable/"bad" characters in strings.
- Fixed the location in a syntax error message.
- Fixed questionable handling of newlines in single-line comment tokens.
Repository:
- Added examples.lua.

v1.2.1 (2021-06-01)
- Fixed lookups and calls that shouldn't parse, like "":sub().

v1.2.0 (2021-05-13)
- Added parser.updateReferences().
- Changed arguments for parser.traverseTree() and the callback.
- Fixed parsing of consecutive unary operations.

v1.1.0 (2020-07-06)
- Added parser.traverseTree().

v1.0.0 (2020-07-05)
- Initial release!


test.lua
-- [===[
local a
local b, c = true, "f\0o"

local _ = 3, 345, 0xff, 0xBEBADA
local _ = 3.0, 3.1416, 314.16e-2, 0.31416E1, 34e1
local _ =  .0,  .1416,    .16e-2,  .31416E1, 34e1
local _ = 0x0.1E, 0xA23p-4, 0X1.921FB54442D18P+1
local _ = "'\'", [[\]]

-- Comment
--[[ Long ]] --[==[
comment.
]==]

do
	local x = 1
	local y = 1.
	local z = .1
	local n = 0x0.8p-1

	local y = (a:b[[]])
	local s = "\0\1\2\3\4\5\6\7\8\9\10\11\12\13\\\01499\15\16\17\18\19\"\20\21\22\23¤\24\25\26\27\28\29\30\31~ëÓ§\255`\0011\1\z

	\xff\u{b4}"

	local x = 1 + 2 + 3
	local y = (1 - 2) - 3
	local z = 1 - (2 - 3)

	do foo() ; o:m() end

	while false or true do
		foo()
	end
	repeat
		local x = x + foo()
	until x - -7 == 8

	for i = 1, 9 do  foo(i)  end
	for i, v in ipairs(t) do  bar(i, v)  end

	local function foo() end
	local function foo(a, ...) return end
	function a:b(x, ...) return ... end
	a[x] = function(i, j) if i then return end return 5 end
	g = a[x]()

	local a = 1 + 2 + 3
	local b = "1".."2".."3"
	local c = 1 + 4^3^2 * 5

	local s = a   .. b
	local s = 1.  .. x
	local s = 1.1 .. x

	local x = a and b and c or q and w or 5

	t.k1.k2       = x
	;(t.k1).k2    = x
	;(t1 or t2).k = x

	;(t1 or t2).k, (t3 or t4).k = x, y

	if x or y then foo() end

	if x then
		foo()
	elseif y then
		bar()
	else
		baz()
	end

	local t = {x=5, ["y"]=(a or b), f=function(...)return"bark"end}

	local m = -8
	local n = not false or false
	local l = #t + 1
	local b = not not not true

	local s1 = "\"''a''b''c''\""
	local s2 = '\'""a""b""c""\''

	print(a .. ... .. 3.75 .. 400)

	local s = (""):rep(9)
	local n = ({9})[1]
	local v = (...)[1]
	local b = (function(n) return n>1 end)(9)

	;;;;

	-- Special number notations.
	local n = - -5
	local n = -1/0
	local n = - -1 / 0
	local n = - (-1/0)
	local n = 5 - 1/0
	local n = - - (-0)

	-- [[ Lua 5.2+
	goto foo
	do goto foo end
	::foo::

	local a = x & y
	local b = x ~ y
	local c = x | y
	local d = x << y
	local e = x >> y
	local f = x // y
	local g = ~x

	local foo = ((y & 0x7) << 4) | 0xFF00

	while 1 do
		break
		unreachable()
	end
	--]]

	-- [[ Lua 5.4+
	local x<close>
	local y<close>, z<const> = 1, 2
	--]]
end

-- Minify tests.
do
	local longNameA = global    -- Watchers: 3
	local longNameB = longNameA -- Watchers: 5
	local global    = global    -- Watchers: 4
	local longNameC = longNameB -- Watchers: 3
	local getGlobal = e         -- Watchers: 2
	local longNameD = longNameB -- Watchers: 1
	local longNameE = longNameA -- Watchers: 0

	function globalFunc(x) local x = x end

	local upvalue = 0

	local function f(arg1, arg2)
		local assign1, assign2 = 0, 0
		local decl, func

		decl = 0
		func = function()end

		for loop1, loop2 in ipairs(t) do
			func(upvalue, arg1,arg2, assign1,assign2, decl, loop1,loop2)
		end
	end

	-- [[ Lua 5.2+
	local _ENV = _ENV
	local function getEnv()
		return _ENV
	end
	--]]
end
--]===]

return


examples.lua
--[[============================================================
--=
--=  Examples for Dumb Lua Parser
--=
--=    1 - Find globals
--=    2 - Detect variable shadowing
--=    3 - Replace calls to a function
--=    4 - Construct an AST (the painful way)
--=
--=-------------------------------------------------------------
--=
--=  Dumb Lua Parser - Lua parsing library
--=  by Marcus 'ReFreezed' Thunström
--=
--============================================================]]

local parser = require("dumbParser")

io.stdout:setvbuf("no")
io.stderr:setvbuf("no")



--
-- 1 - Find globals
--

local ast = parser.parse(
	[[
		local x = math.floor(1.5) -- 'math' is a global.
		y       = "foo"           -- 'y' is a global.
	]],
	"dummyFilePath.lua"
)

parser.updateReferences(ast)
print("Globals:")

-- Note that parser.findGlobalReferences() could be used instead of this.
parser.traverseTree(ast, function(node)
	if node.type == "identifier" and not node.declaration then
		print(parser.formatMessage(node, "Found global '%s'.", node.name))
	end
end)

print("\n\n")



--
-- 2 - Detect variable shadowing
--

local ast = parser.parse[[
	local function foo(x, x)
		for x = x-1, x+1 do
			local x = "foo"
		end
		local x = x
	end
]]

parser.updateReferences(ast)
print("Shadows:")

-- Note that findShadows() doesn't detect shadowing of globals.
for _, shadows in ipairs(parser.findShadows(ast)) do
	for i, ident in ipairs(shadows) do
		-- The first item in 'shadows' is the identifier shadowing the rest.
		local message = (i == 1) and "This identifier is shadowing others:" or "Shadowed #"..i-1
		print(parser.formatMessage(ident, message))
	end
end

print("\n\n")



--
-- 3 - Replace calls to a function
--

--
-- Here we change calls to assert() into 'if' statements:
--
--     -- From this.
--     assert(foo)
--
--     -- Into this.
--     if not (foo) then
--         error("Assertion failed: foo")
--     end
--

local ast = parser.parse[[
	local function calculate(x)
		assert(type(x) == "number")
		-- ...
	end
	local function doThing(foo)
		assert(foo ~= "bar")
		-- ...
	end
]]

parser.traverseTree(ast, function(node, parent, container, key)
	if node.type == "call" and node.callee.type == "identifier" and node.callee.name == "assert" then
		-- We assume the parent is a block, i.e. that the call is a complete statement.
		local oldCall      = node
		local oldCondition = oldCall.arguments[1]

		-- Create the 'if' statement and the new condition.
		local ifNode    = parser.newNode("if")
		ifNode.bodyTrue = parser.newNode("block")

		local condition      = parser.newNode("unary", "not")
		condition.expression = oldCondition
		ifNode.condition     = condition

		-- Create the error() call.
		local message = "Assertion failed: " .. parser.toLua(oldCondition)

		local errorCall               = parser.newNode("call")
		errorCall.callee              = parser.newNode("identifier", "error")
		errorCall.arguments[1]        = parser.newNode("literal", message)
		ifNode.bodyTrue.statements[1] = errorCall

		-- Replace the node.
		container[key] = ifNode

		-- Prevent traversal down through the old call.
		return "ignorechildren"
	end
end)

print("Transformed code:")
print(parser.toLua(ast, true))
print("\n\n")



--
-- 4 - Construct an AST (the painful way)
--

--[[ This is the program we will create:
function _G.double(n)
	return 2 * n
end
local x = double(5)
]]

local new = parser.newNode

local block = new("block")

local assignment             = new("assignment")
assignment.targets[1]        = new("lookup")
assignment.targets[1].object = new("identifier", "_G")
assignment.targets[1].member = new("literal", "double")
table.insert(block.statements, assignment)

local func           = new("function")
func.parameters[1]   = new("identifier", "n")
func.body            = new("block")
assignment.values[1] = func

local ret           = new("return")
ret.values[1]       = new("binary", "*")
ret.values[1].left  = new("literal", 2)
ret.values[1].right = new("identifier", "n")
table.insert(func.body.statements, ret)

local decl                  = new("declaration")
decl.names[1]               = new("identifier", "x")
decl.values[1]              = new("call")
decl.values[1].callee       = new("identifier", "double")
decl.values[1].arguments[1] = new("literal", 5)
table.insert(block.statements, decl)

print("Constructed code:")
print(parser.toLua(block, true))
print("\n\n")




runTest.lua
-- Parsing tests.
local PRETTY_OUTPUT   = 1==1
local PRINT_IDS       = 1==1
local PRINT_LOCATIONS = 1==1

io.stdout:setvbuf("no")
io.stderr:setvbuf("no")

local testCount = 0
local failCount = 0
local results   = {}

local loadLuaString   = loadstring or load
local parser          = require"dumbParser"
parser.printIds       = PRINT_IDS
parser.printLocations = PRINT_LOCATIONS

local function test(label, f)
	testCount = testCount + 1

	print("Running test: "..label)
	local ok, err = pcall(f)

	if ok then
		print("Test succeeded!")
	else
		failCount = failCount + 1
		io.stderr:write(err, "\n")
		print("Test failed!")
	end

	table.insert(results, {label=label, ok=ok, err=err})
end

local function assert(v, err)
	if v then  return v  end
	err = tostring(err or "Assertion failed!")
	error(debug.traceback(err, 2), (err:find"^Error @" and 0 or 2))
end

local function assertLua(lua, expectedLua, level)
	expectedLua = expectedLua:gsub("^%s+", ""):gsub("%s+$", "")
	if lua == expectedLua then  return  end

	error(string.format(
		"Unexpected Lua output.\n%s\nExpected:\n%s\n%s\nGot:\n%s\n%s",
		("-"):rep(64),
		expectedLua,
		("-"):rep(64),
		lua,
		("-"):rep(64)
	), 1+(level or 1))
end

test("Test file", function()
	local path = "test.lua"
	-- local path = "dumbParser.lua"

	-- assert(loadfile(path))

	local tokens = assert(parser.tokenizeFile(path))
	local ast    = assert(parser.parse(tokens))

	-- parser.printTokens(tokens)
	-- parser.printTree(ast)
	-- error("DEBUG")

	--[[
	parser.traverseTree(ast, function(node, parent, container, k)
		io.write(tostring(container), ".", tostring(k), " ")
		parser.printNode(node)
		-- if container then
		-- 	container[k]        = parser.newNode("call")
		-- 	container[k].callee = parser.newNode("identifier", "foo")
		-- end
	end)
	--]]

	parser.traverseTree(ast, function(node)
		if node.type == "identifier" and node.name == "foo" then
			node.name = "dog"
		end
	end)

	parser.updateReferences(ast)
	-- parser.printTree(ast)
	-- print(parser.toLua(ast, 1==1))
	-- error("DEBUG")

	if 1==1 then
		parser.simplify(ast)
		parser.clean(ast)
		parser.simplify(ast)
	else
		parser.minify(ast)
	end
	-- parser.printTree(ast)
	-- print(parser.toLua(ast, 1==1))
	-- error("DEBUG")

	local lua = assert(parser.toLua(ast, PRETTY_OUTPUT))

	do
		local luaEdit = lua
		-- luaEdit = luaEdit:gsub(("."):rep(250), "%0\0")
		-- luaEdit = luaEdit:gsub("([%w_]+)%z([%w_]+)", "%1%2\n")
		-- luaEdit = luaEdit:gsub("%z", "\n")
		print(luaEdit)
	end

	assert(loadLuaString(lua, "@<luastring>"))

	-- Round-trip.
	local tripTokens = assert(parser.tokenize(lua))
	local tripAst    = assert(parser.parse(tripTokens))
	local tripLua    = assert(parser.toLua(tripAst, PRETTY_OUTPUT))

	if tripLua ~= lua then
		print(("-"):rep(64))
		print(lua)
		print(("-"):rep(64))
		print(tripLua)
		print(("-"):rep(64))
		error("Failed round-trip.")
	end
end)

test("Token stream manipulations", function()
	local tokens = parser.newTokenStream()

	parser.insertToken(tokens, "identifier",  "math")
	parser.insertToken(tokens, "punctuation", ".")
	parser.insertToken(tokens, "identifier",  "abs")
	parser.insertToken(tokens, "punctuation", "(")
	parser.insertToken(tokens, "punctuation", "-")
	parser.insertToken(tokens, "number",      1.75)
	parser.insertToken(tokens, "punctuation", ")")

	parser.insertToken(tokens, 1, "keyword",     "local")
	parser.insertToken(tokens, 2, "identifier",  "n")
	parser.insertToken(tokens, 3, "punctuation", "=")
	parser.insertToken(tokens, 4, "punctuation", "/")

	parser.removeToken(tokens, 4)

	local ast = assert(parser.parse(tokens))
	assertLua(assert(parser.toLua(ast)), [[ local n=math.abs(-1.75); ]])
end)

test("AST manipulations", function()
	local identifier1  = parser.newNode("identifier", "foo")
	local identifier2  = parser.newNode("identifier", "foo", "const")
	local vararg       = parser.newNode("vararg")
	local literal1     = parser.newNode("literal", 8.125)
	local literal2     = parser.newNode("literal", "foo")
	local literal3     = parser.newNode("literal", true)
	local literal4     = parser.newNode("literal", nil)
	local tableNode    = parser.newNode("table")
	local lookup       = parser.newNode("lookup")
	local unary1       = parser.newNode("unary",  "#")
	local unary2       = parser.newNode("unary",  "not")
	local binary1      = parser.newNode("binary", "/")
	local binary2      = parser.newNode("binary", "<<")
	local call         = parser.newNode("call")
	local functionNode = parser.newNode("function")
	local breakNode    = parser.newNode("break")
	local returnNode   = parser.newNode("return")
	local label        = parser.newNode("label", "foo")
	local gotoNode     = parser.newNode("goto",  "foo")
	local block        = parser.newNode("block")
	local declaration  = parser.newNode("declaration")
	local assignment   = parser.newNode("assignment")
	local ifNode       = parser.newNode("if")
	local whileLoop    = parser.newNode("while")
	local repeatLoop   = parser.newNode("repeat")
	local forLoop      = parser.newNode("for", "numeric")

	do
		parser.setChild(call,        "callee", identifier1)
		parser.setChild(declaration, "names",  1, identifier2)

		parser.addChild(tableNode, "fields", literal2, literal1)

		parser.printNode(parser.getChild(call,        "callee"))
		parser.printNode(parser.getChild(declaration, "names", 1))
		parser.printNode(parser.getChild(tableNode,   "fields", 1, "value"))

		parser.removeChild(tableNode, "fields", 1)
	end

	local block = assert(parser.parse([[
		local x = 49
	]], "<luastring>"))

	assert(block.type == "block")
	assert(block.statements[1])
	assert(block.statements[1].type == "declaration")
	assert(block.statements[1].names[1])
	assert(block.statements[1].names[1].type == "identifier")
	assert(block.statements[1].names[1].name == "x")
	assert(block.statements[1].values[1])
	assert(block.statements[1].values[1].type  == "literal")
	assert(block.statements[1].values[1].value == 49)

	block.statements[1].names[1]  = identifier1
	block.statements[1].values[1] = literal1

	local lua = assert(parser.toLua(block, PRETTY_OUTPUT))
	print(lua)

	assert(loadLuaString(lua, "@<luastring>"))
end)

test("Call and lookup parsing", function()
	assert(    parser.parse([[ x =  tbl   () ]], "<luastring>"))
	assert(    parser.parse([[ x =  tbl   "" ]], "<luastring>"))
	assert(    parser.parse([[ x =  tbl .m() ]], "<luastring>"))
	assert(    parser.parse([[ x =  tbl :m"" ]], "<luastring>"))
	assert(    parser.parse([[ x = (tbl)  () ]], "<luastring>"))
	assert(    parser.parse([[ x = (tbl)  "" ]], "<luastring>"))
	assert(    parser.parse([[ x = (tbl).m() ]], "<luastring>"))
	assert(    parser.parse([[ x = (tbl):m"" ]], "<luastring>"))
	assert(not parser.parse([[ x =  " "   () ]], "<luastring>"))
	assert(not parser.parse([[ x =  " "   "" ]], "<luastring>"))
	assert(not parser.parse([[ x =  " " .m() ]], "<luastring>"))
	assert(not parser.parse([[ x =  " " :m"" ]], "<luastring>"))
	assert(    parser.parse([[ x = (" ")  () ]], "<luastring>"))
	assert(    parser.parse([[ x = (" ")  "" ]], "<luastring>"))
	assert(    parser.parse([[ x = (" ").m() ]], "<luastring>"))
	assert(    parser.parse([[ x = (" "):m"" ]], "<luastring>"))
	assert(not parser.parse([[ x =  { }   () ]], "<luastring>"))
	assert(not parser.parse([[ x =  { }   "" ]], "<luastring>"))
	assert(not parser.parse([[ x =  { } .m() ]], "<luastring>"))
	assert(not parser.parse([[ x =  { } :m"" ]], "<luastring>"))
	assert(    parser.parse([[ x = ({ })  () ]], "<luastring>"))
	assert(    parser.parse([[ x = ({ })  "" ]], "<luastring>"))
	assert(    parser.parse([[ x = ({ }).m() ]], "<luastring>"))
	assert(    parser.parse([[ x = ({ }):m"" ]], "<luastring>"))
	assert(not parser.parse([[ x =  123   () ]], "<luastring>"))
	assert(not parser.parse([[ x =  123   "" ]], "<luastring>"))
	assert(not parser.parse([[ x =  123 .m() ]], "<luastring>"))
	assert(not parser.parse([[ x =  123 :m"" ]], "<luastring>"))
	assert(    parser.parse([[ x = (123)  () ]], "<luastring>"))
	assert(    parser.parse([[ x = (123)  "" ]], "<luastring>"))
	assert(    parser.parse([[ x = (123).m() ]], "<luastring>"))
	assert(    parser.parse([[ x = (123):m"" ]], "<luastring>"))
	assert(not parser.parse([[ x =  nil   () ]], "<luastring>"))
	assert(not parser.parse([[ x =  nil   "" ]], "<luastring>"))
	assert(not parser.parse([[ x =  nil .m() ]], "<luastring>"))
	assert(not parser.parse([[ x =  nil :m"" ]], "<luastring>"))
	assert(    parser.parse([[ x = (nil)  () ]], "<luastring>"))
	assert(    parser.parse([[ x = (nil)  "" ]], "<luastring>"))
	assert(    parser.parse([[ x = (nil).m() ]], "<luastring>"))
	assert(    parser.parse([[ x = (nil):m"" ]], "<luastring>"))
end)

test("Simplify", function()
	local function testSimplify(lua, expectedLua)
		local ast = assert(parser.parse(lua, "<luastring>"))
		parser.simplify(ast)
		assertLua(assert(parser.toLua(ast)), expectedLua, 2)
	end

	testSimplify([[ x = 1+2   ]], [[ x=3; ]])
	testSimplify([[ x = 1+2^3 ]], [[ x=9; ]])

	testSimplify([[ x = 1 << 8     ]], [[ x=256;   ]])
	testSimplify([[ x = 2^99999    ]], [[ x=(1/0); ]])
	testSimplify([[ x = 5 - - - -5 ]], [[ x=10;    ]])

	testSimplify([[ x = 2779          ]], [[ x=2779;  ]])
	testSimplify([[ x = ~2779         ]], [[ x=-2780; ]])
	testSimplify([[ x = 2779 & 0x1011 ]], [[ x=17;    ]])
	testSimplify([[ x = 2779 ~ 0x1011 ]], [[ x=6858;  ]])
	testSimplify([[ x = 2779 | 0x1011 ]], [[ x=6875;  ]])

	testSimplify([[ x =  2       ]], [[ x=2;  ]])
	testSimplify([[ x =  2 >>  1 ]], [[ x=1;  ]])
	testSimplify([[ x =  2 <<  1 ]], [[ x=4;  ]])
	testSimplify([[ x = -2       ]], [[ x=-2; ]])
	testSimplify([[ x = -2 >>  1 ]], (parser.INT_SIZE == 32 and [[ x=2147483647; ]] or [[ x=9223372036854775807; ]]))
	testSimplify([[ x = -2 <<  1 ]], [[ x=-4; ]])
	testSimplify([[ x =  2 >>  0 ]], [[ x=2;  ]])
	testSimplify([[ x =  2 <<  0 ]], [[ x=2;  ]])
	testSimplify([[ x = -2 << -1 ]], (parser.INT_SIZE == 32 and [[ x=2147483647; ]] or [[ x=9223372036854775807; ]]))
	testSimplify([[ x = -2 >> -1 ]], [[ x=-4; ]])

	testSimplify([[ x = 1/0 & 2 ]], [[ x=(1/0)&2; ]])

	testSimplify([[ x = "yes" == "no" ]], [[ x=false; ]])
	testSimplify([[ x = 5     ~= nil  ]], [[ x=true;  ]])
	testSimplify([[ x = 80.6  >= 34   ]], [[ x=true;  ]])

	testSimplify([[ x = "one" .. 2 .. "three" ]], [[ x="one2three"; ]])

	testSimplify([[ x = true  and always1() ]], [[ x=always1(); ]])
	testSimplify([[ x = true  or  never1()  ]], [[ x=true;      ]])
	testSimplify([[ x = 0     and always2() ]], [[ x=always2(); ]])
	testSimplify([[ x = 0     or  never2()  ]], [[ x=0;         ]])
	testSimplify([[ x = false and never3()  ]], [[ x=false;     ]])
	testSimplify([[ x = false or  always3() ]], [[ x=always3(); ]])
	testSimplify([[ x = nil   and never4()  ]], [[ x=nil;       ]])
	testSimplify([[ x = nil   or  always4() ]], [[ x=always4(); ]])

	testSimplify(
		[[
		if 9 > 1 then
			ifYes()
			if 1 == "foo" then  ifNo()  end
		end
		]],
		[[ ifYes(); ]]
	)
	testSimplify(
		[[
		while 9 > 1 do
			whileYes()
			while 1 == "foo" do  whileNo()  end
		end
		]],
		[[ while true do whileYes();end ]]
	)
	testSimplify(
		[[
		repeat
			repeatYes()
			repeat  repeatOnce()  until 1 ~= "foo"
		until 9 <= 1
		]],
		[[ repeat repeatYes();repeatOnce();until false ]]
	)
end)

print(("="):rep(64))
for _, result in ipairs(results) do
	print(string.format("%-4s  %s", (result.ok and "ok" or "FAIL"), result.label))
end
print(("="):rep(64))

if failCount > 0 then
	print("Error: "..failCount.."/"..testCount.." tests failed!")
	os.exit(1)
else
	print("All tests passed successfully!")
end
